noir-projects/aztec-nr/aztec/src/hash.nr

```diff
@@ -110,13 +116,36 @@ pub fn hash_calldata_array<let N: u32>(calldata: [Field; N]) -> Field {
 
 // Same as `hash_calldata_array`, but takes a slice instead of an array.
 pub fn hash_calldata(calldata: [Field]) -> Field {
+    // Safety: this is only called to detect calls that will result in transactions rejected by the nodes in order to
+    // help developers prevent them. If a call with too much calldata was made, then the resulting transaction would
+    // simply be ignored by most nodes in the network, and likely never included in a block.
+    unsafe { check_calldata_len(calldata); };
+
     if calldata.len() == 0 {
         0
     } else {
         poseidon2_hash_with_separator_slice(calldata, GENERATOR_INDEX__PUBLIC_CALLDATA)
     }
 }
 
+unconstrained fn check_calldata_array_len<let N: u32>(calldata: [Field; N]) {
+    check_calldata_len_inner(calldata.len());
+}
+
+unconstrained fn check_calldata_len(calldata: [Field]) {
+    check_calldata_len_inner(calldata.len());
+}
+
+unconstrained fn check_calldata_len_inner(len: u32) {
+    // Technically what's restricted is the length of the sum of the calldata for all enqueued functions. For simplicity
+    // we restrict each enqueued call to not exceed their fair share.
+    let max = (MAX_FR_CALLDATA_TO_ALL_ENQUEUED_CALLS / MAX_ENQUEUED_CALLS_PER_CALL);
```

Comment:
You may want to get Mike's eyes on this. We explicitly chose the limit so that it applies to the sum because doing what is done here might be too restrictive (for some L3 ideas that were around, etc)

Reply:
Assuming that people simulate before sending maybe there is some other place to do this in TS where all the calldatas are available? (not sure)

Reply:
Ah, will change the approach then. I was just going off of David's comment there. Yes, we can look at the object we produce prior to broadcast, prior to proving even, this would work even if we don't do any public simulation.